src/components/createContainer.jsx

```diff
@@ -0,0 +1,108 @@
+/* @flow */
+
+import React, { Component, PropTypes } from 'react';
+import invariant from 'invariant';
+
+import getDisplayName from '../utils/getDisplayName';
+import createAdaptorShape from '../utils/createAdaptorShape';
+
+const adaptorShape = createAdaptorShape(PropTypes);
+
+export default function createContainer(DecoratedComponent, specs) {
+  const displayName = `AdrenalineContainer(${getDisplayName(DecoratedComponent)})`;
+
+  invariant(
+    specs !== null && specs !== undefined,
+    `${displayName} requires configuration.`
+  );
+  invariant(
+    typeof specs.queries === 'function',
+    `You have to define 'queries' as a function.`
+  );
+
+  return class extends Component {
+    static displayName = displayName
+    static DecoratedComponent = DecoratedComponent
+
+    static contextTypes = {
+      renderLoading: PropTypes.func.isRequired,
+      adaptor: adaptorShape.isRequired,
+    }
+
+    constructor(props, context) {
+      super(props, context);
+
+      const { adaptor } = context;
+      //DecoratedComponent.prototype.shouldComponentUpdate = adaptor.shouldComponentUpdate;
+
+      const initialArgs = specs.args(props);
+
+      this.state = {
+        data: {},
+        args: specs.args(props),
+      };
+
+      const dataKeys = Object.keys(specs.queries(initialArgs));
+      const { propTypes } = DecoratedComponent;
+      const requiredKeys = dataKeys.filter(key =>
+        propTypes[key] && !propTypes[key].isRequired
+      );
+
+      this.isDataLoaded = (data = {}) => requiredKeys.every(key => {
```

Comment:
unfortunately, a loading state is more nuanced, considering that arg changes can force new queries. But ignoring that, this makes it impossible to intentionally resolve 'null' or 'undefined'. Since fetch/resolve is promise based, you have information about whether an uncommitted slice is resolving, and you could approach it that way.